Koha/ApiKey.pm

```diff
@@ -0,0 +1,97 @@
+package Koha::ApiKey;
+
+# Copyright BibLibre 2015
+#
+# This file is part of Koha.
+#
+# Koha is free software; you can redistribute it and/or modify it under the
+# terms of the GNU General Public License as published by the Free Software
+# Foundation; either version 3 of the License, or (at your option) any later
+# version.
+#
+# Koha is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with Koha; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+use Modern::Perl;
+
+use Carp;
+
+use Koha::Database;
+use Koha::Exceptions;
+
+use UUID;
+
+use base qw(Koha::Object);
+
+=head1 NAME
+
+Koha::ApiKey - Koha API Key Object class
+
+=head1 API
+
+=head2 Class methods
+
+=head3 store
+
+    my $api_key = Koha::ApiKey->new({ patron_id => $patron_id })->store;
```

Comment:
Hmmm, so we're tieing a 'client' directly to a 'patron', this isn't how OAuth usually works and could cause some fun down the line when attempting to add the different grant types.  I would suggest instead of tying to a user at this point, we add a 'list of permissions' interface and a table for setting privileges at the client level.

When it comes to implementing other grant flows later this will work nicely as it allows for different clients to be given different 'maximal' permissions which the user can then reduce during the grant flow themselves to just what they are happy giving the client.

I'm wondering why 'Client Credentials Grant' was chosen as the first grant type as it's usually recommended against for security reasons ;)

Reply:
You are right about the maximal permissions. But we still need an extra layer in the backend, to make decisions like 'what patrons the client has permissions to edit'

Client credentials is required by several potential consumers (major ones). We are not in a position to tell them to not require this grant, and there's no reason not to implement it, and we already did it :-D. I agree it doesn't make sense to allow OPAC users give third party consumers access to Koha using this grant like this. I resurrected this patches from the original bug and it felt like having an on/off switch (off by default) was a 'good enough' measure.